lib/Conversion/TorchToMhlo/ViewLikeOps.cpp

```diff
@@ -0,0 +1,132 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir/Conversion/TorchToMhlo/TorchToMhlo.h"
+
+#include "../PassDetail.h"
+#include "./PopulatePatterns.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir-hlo/Dialect/mhlo/IR/chlo_ops.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "torch-mlir/Conversion/Utils/Utils.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchDialect.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/TorchUpstream.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionDialect.h"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionOps.h"
+#include <numeric>
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+using namespace mlir::torch::TorchConversion;
+
+
+namespace {
+
+// This defines a template to construct ops whose legalizations are
+// specialized.
+template <typename AtenOpT>
+class ConvertAtenViewOp : public OpConversionPattern<AtenOpT> {
+ public:
+  using OpConversionPattern<AtenOpT>::OpConversionPattern;
+  using OpAdaptor = typename AtenOpT::Adaptor;
+
+  LogicalResult matchAndRewrite(
+      AtenOpT op,
+      OpAdaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    auto rankType =
+        adaptor.self().getType().template dyn_cast<RankedTensorType>();
+    if (!rankType)
+      return op.emitError("Only ranked tensor types are currently supported");
+
+    SmallVector<Value, 4> dimSizes;
+    if (!getAtenViewOpSizes(op, adaptor, rewriter, dimSizes)) {
+      return op.emitError("Dims size must be a list of Scalar");
+    }
+
+    auto loc = op.getLoc();
+    auto newRank = dimSizes.size();
+    if (newRank == 0 || rankType.getRank() == 0) {
+      rewriter.replaceOpWithNewOp<mhlo::ReshapeOp>(
+          op,
+          OpConversionPattern<AtenOpT>::getTypeConverter()->convertType(
+              op.getType()),
+          adaptor.self());
+      return success();
+    }
+
+    std::for_each(dimSizes.begin(), dimSizes.end(), [&](Value& dSize) {
+      dSize = rewriter.create<ToI64Op>(loc, dSize).getResult();
+      return dSize;
+    });
+
+#ifdef TORCH_MLIR_ENABLE_MHLO_TRUNC_DIMSIZE_TO_I32
+    // The i64 calculation is much slower than i32 on some devices, such as Nvidia GPU.
+    // One can truncate from i64 to i32 since dimension sizes are unlikely to exceed
+    // the range of i32(4GiB)
+    std::for_each(dimSizes.begin(), dimSizes.end(), [&](Value& dSize) {
+      // dimSize: cast i64 -> i32
+      dSize = rewriter.create<arith::TruncIOp>(loc, rewriter.getI32Type(), dSize);
+      return dSize;
+    });
+#endif
+
+    Value mhloShape = rewriter.create<tensor::FromElementsOp>(loc, dimSizes);
+    rewriter.replaceOpWithNewOp<chlo::DynamicReshapeOp>(
```

Comment:
Can we use mhlo::DynamicReshapeOp?

Reply:
I think we should not use [`mhlo::DynamicReshapeOp`](https://github.com/tensorflow/mlir-hlo/blob/master/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td#L1900); because it is not sufficient to represent the reshape from Torch Dialect, since a single new dimension specified as -1 is allowed(the size -1 will be inferred from other dimensions)

[`chlo::DynamicReshapeOp`](https://github.com/tensorflow/mlir-hlo/blob/master/include/mlir-hlo/Dialect/mhlo/IR/chlo_ops.td#L909) is almost the same as  [`mhlo::DynamicReshapeOp`](https://github.com/tensorflow/mlir-hlo/blob/master/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.td#L1900) , but it allows a single dimension to be specified as -1, and it will be computed by the op to create a valid reshape.

Reply:
As an exchangeable IR format, It would be good if we can converge the produced result into a single layer, so that the backend do not need to handle two set of dialects. Could you try to:
- represent the rewriting with pure mhlo like here: https://github.com/tensorflow/mlir-hlo/blob/master/lib/Dialect/mhlo/transforms/chlo_legalize_to_hlo.cc#L1698-L1707  
- Or add a chlo_legalize_to_hlo transform into the TorchToMhlo pipeline?

Reply:
Thanks for advice, I had eliminated the usages of chlo dialect.